layers/shell/packages.el

```diff
@@ -105,6 +107,19 @@ is achieved by adding the relevant text properties."
                     'spacemacs//eshell-auto-end nil t))
         (when (configuration-layer/package-usedp 'semantic)
           (semantic-mode -1)))
+
+      ;; Defining a function like this makes it possible to type 'clear' in eshell and have it work
+      (defun eshell/clear ()
+        (interactive)
+        (let ((inhibit-read-only t))
+          (erase-buffer))
+        (eshell-send-input))
+
+      ;; Caution! this will erase buffer's content at C-l
+      (add-hook 'eshell-mode-hook
+         #'(lambda ()
+             (define-key eshell-mode-map (kbd "C-l") 'eshell/clear)
```

Comment:
https://gitter.im/syl20bnr/spacemacs?at=567b4a224f069158055bab7a

Eshell redefines `eshell-mode-map` from scratch in every new eshell buffer.